### PR TITLE
greylisting - finished - running live test

### LINKED ISSUE #1110: greylisting plugin
https://github.com/celesteking/Haraka/pull/1

Can't diff against your repo because apparently my tracking is f.. up.

### PR DESCRIPTION
**Pull request by Dexus**
close #1110 orgin: celesteking/Haraka/pull/1

Plugin seems to consider only IPv4. IPv6 review shall be conducted.

* clean up
* one config file and not 4...
* remove "special-sender"
* greylist is greylist and dont need reject/learn actions

## Comments from commit promoted here by @msimerson

*Fix IPv6 incompatibility
* Update docs / config info for action "learn".
* rework: remove Q deps
* rework: async - done
* set "return" to prevent double callbacks
* fix bug cb not a function

### CODE REVIEW

**Comment by Dexus** [comment]
@smfreegard 
Have restored with IPv6 compatibility. 

>from: celesteking/Haraka/pull/1#issuecomment-132935885
>
>Initial comments after a quick scan:
>
> - LOL - someone read my greylisting paper ;-).
> - actions are undocumented. I'm a bit fuzzy on how 'learn' can possibly work (and be remotely accurate) and how 'black' isn't crazy dangerous.
> - 'special cases' are undocumented?

I don't have looked for the undocumented actions in detail. But __learn__ action will collect data but not reject or defer mails.

The 'special cases' i don't know. But it looks as if dynamic rDNA of certain hosters, only the IP should use instead of the domain.

**Comment by Dexus** [comment]
[ ] remove Q and use call backs.

**Comment by Dexus** [comment]
Works on live system like a busy bee :)
Last hours 29k greylisted; 100 whitelisted - works.
Now I'll be watching the next week.

File: tests/plugins/data.greylist.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            "ip":[
+                [{"octets":[123,123,123,234]},32],
+                [{"parts":[10754,33284,54784,32864,31008,16448,8430,38528]},128],
+                [{"octets":[123,210,123,234]},27],
+                [{"parts":[10754,33284,54784,32864,31008,61166,0,0]},96]
+            ]
+        };
+        this.plugin.list = {"dyndom":["sgvps.net"]};
+        test.expect(7);
+        test.equal(true,  this.plugin.addr_in_list('mail', 'josef@examlpe.com'));
```
Comment (msimerson) [inline tests/plugins/data.greylist.js]: Ahah, I think I see your problem.

````
this.plugin.whitelist = {
+            "mail":{"josef@examlpe.com":true},
````

You are setting this.plugin.whitelist.mail above.

Here you are accessing this.plugin.mail.

**Comment by smfreegard** [comment]
So the 'learn' mode is staying?   
What constitutes a 'special-sender' and why would you want to defer the deferral until hook_data?

**Comment by msimerson** [comment]
> So the 'learn' mode is staying?

I sure hope so. IMO, the only sensible way to deploy greylisting is to run it in learn mode for a few weeks before deploying.

**Comment by Dexus** [comment]
@smfreegard 
special-sender are (postmaster|double-bounce(\-?\d{8,12})?) they never should be greylisted.
i call that now on rcpt_ok

**Comment by smfreegard** [comment]
>> So the 'learn' mode is staying?

> I sure hope so. IMO, the only sensible way to deploy greylisting is to run it in learn mode for a few weeks before deploying.

Good luck with that.   You might as well not bother running greylisting at all if you do that.

It's *impossible* to 'learn' in this case.   You can't determine if a remote system implements a retry queue without testing and observing it.

'Learn' in this context basically means, whitelist *everything* that comes along.  Which means that when you do switch this into 'defer' mode - it will do nothing, because everything was whitelisted whether it queued or not.

**Comment by smfreegard** [comment]
> special-sender are (postmaster|double-bounce(-?\d{8,12})?) they never should be greylisted.

Why not?   A sending host either implements a retry queue or it doesn't - the actual sender is irrelevant to that fact.

**Comment by Dexus** [comment]
So now i removed special-sender, default defer action and clean up hook_data step to hook_rcpt_ok.

**Comment by celesteking** [comment]
>> special-sender are (postmaster|double-bounce(-?\d{8,12})?) they never should be greylisted.

>Why not? A sending host either implements a retry queue or it doesn't - the actual sender is irrelevant to that fact.

That's for cases when sender should not be deferred in RCPT TO phase. For ex., email address verification. Returning temp error will probably count as error for them, thus the need for additional code.

**Comment by celesteking** [comment]
> 'Learn' in this context basically means, whitelist everything that comes along

Yes, learning is broken in this regard, sadly. I'm busy working on other things, so I guess no learn mode in near future. On bright side, it won't do much harm to HAM messages, you'll only end up with a bunch of whitelisted spammy hosts that will expire in a month.

**Pull request merged**